tests/daemon/test-circular-file.c

```diff
@@ -609,11 +611,158 @@ test_circular_file_shrink (Fixture      *fixture,
+                                            gconstpointer unused)
+{
+  gssize length = 43;
+  g_autofree gchar *contents = g_malloc0 (length);
+
+  /* To summarize this bug: the metadata file was 43 NUL bytes. The initial
+   * metadata keyfile, if properly written, is 43 bytes long. This is believed
+   * to be caused by g_file_set_contents() not fsync()ing the temporary file
+   * when the target file does not yet exist; poorly-documented semantics
+   * ensue.
+   *
+   * Regardless of cause, a reasonable recovery path is to re-initialize the
+   * metadata, assuming the file is empty.
+   */
+  g_test_bug ("T19953");
+
+  test_circular_file_broken_metadata_file (fixture, contents, length,
+                                           G_KEY_FILE_ERROR_GROUP_NOT_FOUND);
+}
+
+static void
+test_circular_file_metadata_file_junk (Fixture      *fixture,
+                                       gconstpointer unused)
+{
+  gssize length = 43;
+  gchar contents[length];
+
+  /* Totally broken! */
+  memset (contents, '!', length);
+
+  test_circular_file_broken_metadata_file (fixture, contents, length,
+                                           G_KEY_FILE_ERROR_PARSE);
+}
+
+static void
+test_circular_file_metadata_file_missing_max_size (Fixture      *fixture,
+                                                   gconstpointer unused)
+{
+  const gchar contents[] = "[metadata]\nsize=1024\nhead=0\n";
+  const gssize length = sizeof contents;
+
+  /* max_size is missing; just assume the file is empty, since this will never
+   * happen in practice without other keys also being missing unless someone
+   * edits the file by hand and then you get to keep both pieces.
+   */
+  test_circular_file_broken_metadata_file (fixture, contents, length,
+                                           G_KEY_FILE_ERROR_KEY_NOT_FOUND);
+}
+
+static void
+test_circular_file_metadata_file_missing_size (Fixture      *fixture,
+                                               gconstpointer unused)
+{
+  const gchar contents[] = "[metadata]\nmax_size=1024\nhead=27\n";
+  const gssize length = sizeof contents;
+
+  /* size is missing; we should recover by treating the file as empty because
+   * we do not know at what byte we should loop around.
+   */
+  test_circular_file_broken_metadata_file (fixture, contents, length,
+                                           G_KEY_FILE_ERROR_KEY_NOT_FOUND);
+}
+
+static void
+test_circular_file_metadata_file_missing_head (Fixture      *fixture,
+                                               gconstpointer unused)
+{
+  const gchar contents[] = "[metadata]\nmax_size=1024\nsize=1024\n";
+  const gssize length = sizeof contents;
+
+  /* size is missing; we should recover by treating the file as empty, because
+   * we don't know where the head was supposed to be.
+   */
+  test_circular_file_broken_metadata_file (fixture, contents, length,
+                                           G_KEY_FILE_ERROR_KEY_NOT_FOUND);
```

Comment:
And here as well. Actually, it looks like you might need to refactor these tests a bit to accomodate the different way of dealing with missing groups/keys, because the `expected_error` parameter seems like not very useful except for the "junk data" case :-)

Reply:
Right – I decided that this case and the ones below, when the file's actively malformed (rather than being empty or all `NUL` which is "morally empty"), are serious enough that we do still want to return an error, and have the daemon report an event and explicitly reinitialize. So I split this up.

Reply:
Sounds good to me